Core: Allow a teardown function to be returned from `renderToDOM`

tmeasday opened the pull request:
Issue: Stories (both in Canvas and when rendered in `modernInlineRender`) had no way to be properly torn down, leading to memory/event listener leaks when browsing around (as noticed in Angular by @kroeder).

## What I did

Added some stories to highlight the problem to `react-ts` (not sure this is where they should go, and if we should have a way to properly test this @shilman).

Notice the following (without the changes to `react/../render.tsx`):

1. Browse to a buttonX story
2. Go to docs tab
3. Notice event listener for Canvas is still active (you will see click handlers fire one more time than the number of rendered stories in the docs page).
4. Go to docs tab for other buttonX story
5. Notice event listeners from previous docs page are not torn down.

The Canvas event listeners eventually do get torn down because we are ultimately rendering a different story to the same DOM element, and React handles this for us.

But
a) it should happen earlier I think and
b) that doesn't happen automatically in Angular as I understand it.

## Solution

I allowed `renderToDOM` to return a `teardown` function that will be called whenever a `StoryRender` is taken off the screen.

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?

See reproduction above. I will also add a bunch of `PreviewWeb` integration tests once we are satisfied this is the right approach.

joshwooding committed:
Move memoizerific to dependencies in theming

joshwooding committed:
Move memoizerific and qs to dependencies in router

joshwooding committed:
Move memoizerific and qs to dependencies in ui

Andarist committed:
Fixed PnP compatibility for bundled ui and router packages

yannbf committed:
chore(CI): reenable cra pnp tests

ndelangen committed:
bundle the syntax highlighter again inside lib/components

yannbf committed:
fix: change the way presets are required

yannbf committed:
telemetry: strip out preset from addon name

also move getStorybookMetadata method to the bottom so it's used after the definition
of computeStorybookMetadata

tmeasday committed:
Allow a teardown function to be returned from `renderToDOM`

ghengeveld committed:
Don't intercept calls inside callbacks, and bubble exceptions inside callbacks up to the parent rather than forwarding it to the next interceptable call

ghengeveld committed:
Show child interactions in log and display errors in the appropriate place

joshwooding committed:
Warn when storyName usage is detected in CSF3 stories

shilman reviewed (approved):
Code LGTM, but we definitely need some tests for this! 👍

kroeder commented:
@tmeasday can I already create a second draft pr with a fix for angular?

ghengeveld committed:
Fix mock data

shilman commented:
@kroeder You should be able to telescope off this PR and create a second PR on top of it

kroeder commented:
Already working on it and it looks promising!

ghengeveld committed:
Fix exception type

ghengeveld committed:
Disable interaction on child interactions

Commit:
Remove duplicate statement

`title: 'DocumentScreen',` appears twice in the component meta.

tmeasday commented:
@kroeder -- discussed with @shilman and a plan to get this into 6.5 with a minimum of disruptive changes would be as follows:

1. Implement the `TeardownRenderToDOM` more or less as done here in the core.
2. Only actually implement a `RenderToDOM` that returns the above in angular, maybe in docs mode only.

React (and likely others) *should* also return a teardown function, at least for docs mode, but we are concerned about making breaking changes especially to the way components/stories are unmounted when changing stories. I would suggest we should be careful for angular in doing that also, which is why I suggest only returning a teardown function in docs mode, at least for now.

As part of the 7.0 cycle we plan to look closely at various existing and closed issues around remounting and get some test in place that'll give us confidence to make this change everywhere.

joshwooding committed:
Core: Fix process is not defined when using components

ghengeveld committed:
Many fixes

ghengeveld committed:
Remove unused eslint directives

ghengeveld committed:
Fix prop forwarding

ghengeveld committed:
Fix back button

ghengeveld committed:
Fix pausedAt indicator

ghengeveld committed:
Fix broken play function

ghengeveld committed:
Value might be null

darleendenno committed:
fix: display skip to sidebar button

ghengeveld committed:
Tweak indicator style

ghengeveld committed:
Fix rendering inconsistency

ghengeveld committed:
Fix zIndex

ghengeveld committed:
Fix tests

ghengeveld committed:
Fix story data

tmeasday committed:
Add integration tests for teardown function

tmeasday commented:
I added some integration tests. We don't currently have a mechanism for full integration tests with the docs stack (i.e. `<Story>` in modern inline mode) but I would like to add it as part of other canvas tests.

ghengeveld committed:
Show pausedAt indicator in story

ghengeveld committed:
Add story book interaction row with parentId

ghengeveld committed:
Better story data

ghengeveld committed:
Better story data

al3xg0 committed:
Update test-runner.md

Make code snippet consistent with the others on the page. Typically yarn is shown before npm, so this change will make the code snippet consistent with the rest.

ghengeveld committed:
Collapse child interactions

shilman commented on addons/docs/src/blocks/Story.tsx:
```diff
@@ -120,6 +120,7 @@ function makeGate(): [Promise<void>, () => void] {
 }
 
 const Story: FunctionComponent<StoryProps> = (props) => {
+  console.log('Story');
```
```suggestion
```

tmeasday committed:
Update addons/docs/src/blocks/Story.tsx

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

ghengeveld committed:
Expand by default

ghengeveld committed:
Tweak anonymous function color

ghengeveld committed:
Update addons/interactions/src/components/Interaction/Interaction.tsx

Co-authored-by: Michael Shilman <shilman@users.noreply.github.com>

shilman commented:
@tmeasday please update snapshots

tmeasday committed:
Update snapshots

tmeasday committed:
Merge remote-tracking branch 'origin/next' into tom/sb-415-return-cleanup-function-from-rendertodom

tmeasday committed:
Use node14 in netlify

shilman committed:
Merge branch 'next' into tom/sb-415-return-cleanup-function-from-rendertodom

shilman reviewed (approved):
LGTM. Let's get this working in 7.0 Angular before deciding whether to patch to 6.5

kroeder commented:
Awesome! Workin on that starting today :) @shilman
